controllers/configurationpolicy_controller.go

```diff
@@ -506,9 +506,29 @@ func (r *ConfigurationPolicyReconciler) getObjectTemplateDetails(
 	return templateObjs, selectedNamespaces, false, nil
 }
 
-func (r *ConfigurationPolicyReconciler) cleanUpChildObjects(plc policyv1.ConfigurationPolicy) []string {
+func (r *ConfigurationPolicyReconciler) cleanUpChildObjects(plc policyv1.ConfigurationPolicy,
+	related []policyv1.RelatedObject,
+) []string {
 	deletionFailures := []string{}
 
+	// When spec is updated and new related objects are created
+	if len(related) != 0 {
+		log.Info("preparing to delete detached objects...")
+
+		oldRelated := plc.Status.RelatedObjects
+		var objShouldRemoved []policyv1.RelatedObject
+
+		for _, oldR := range oldRelated {
+			exist := containRelated(related, oldR)
+
+			if !exist {
+				objShouldRemoved = append(objShouldRemoved, oldR)
+			}
+		}
+
+		plc.Status.RelatedObjects = objShouldRemoved
+	}
+
 	for _, object := range plc.Status.RelatedObjects {
 		// set up client for object deletion
```

Comment:
I'm thinking out loud here in case this is cleaner (it is to me at least)... what do you think of this?

```
func (r *ConfigurationPolicyReconciler) cleanUpChildObjects(plc policyv1.ConfigurationPolicy,
	newRelated []policyv1.RelatedObject,
) []string {
	deletionFailures := []string{}
	objsToDelete := []policyv1.RelatedObject{}
	
    log.Info("preparing to delete detached objects...")

	if len(newRelated) == 0 { // delete all
	    objsToDelete = plc.Status.RelatedObjects
	} else { // When spec is updated and new related objects are created
		oldRelated := plc.Status.RelatedObjects

		for _, oldR := range oldRelated {
			if !containRelated(newRelated, oldR) {
				objsToDelete = append(objsToDelete, oldR)
			}
		}
	}

	for _, object := range objsToDelete {
		// set up client for object deletion
```

Reply:
Actually, the two cases can just work the same, because if the `newRelated` is empty, then `containRelated` will just always be false.

```
func (r *ConfigurationPolicyReconciler) cleanUpChildObjects(plc policyv1.ConfigurationPolicy,
	newRelated []policyv1.RelatedObject,
) []string {
	deletionFailures := []string{}
	objsToDelete := []policyv1.RelatedObject{}
	
    log.Info("preparing to delete detached objects...")
    
	oldRelated := plc.Status.RelatedObjects
	for _, oldR := range oldRelated {
		if !containRelated(newRelated, oldR) {
			objsToDelete = append(objsToDelete, oldR)
		}
	}
		
	for _, object := range objsToDelete {
		// set up client for object deletion
```

Reply:
what is deletionFailures

Reply:
also logic is same but change is variable name right?

Reply:
we shouldn't change `for _, object := range objsToDelete {`

Reply:
when newRelated  = nil we have to fall in loop, unnecessary isn't it?